crates/ln-dlc-node/src/networking/tungstenite.rs

```diff
@@ -0,0 +1,171 @@
+                )
+                .await
+                {
+                    Ok(ControlFlow::Break(())) => break,
+                    Ok(ControlFlow::Continue(())) => (),
+                    Err(err) => {
+                        error!("Disconnecting websocket with error: {err}");
+                        let _ = ws.close(None).await;
+                        peer_manager.as_ref().socket_disconnected(&descriptor);
+                        peer_manager.as_ref().process_events();
+                        break;
+                    }
+                }
+            }
+        })
+    } else {
+        None
+    }
+}
+
+async fn process_messages<PM>(
+    peer_manager: &PM,
+    task_rx: &mut UnboundedReceiver<BgTaskMessage>,
+    ws: &mut WebSocketStream<MaybeTlsStream<TcpStream>>,
+    descriptor: &mut DynamicSocketDescriptor,
+    emit_read_events: &mut bool,
+) -> Result<ControlFlow<()>, anyhow::Error>
+where
+    PM: Deref + 'static + Send + Sync + Clone,
+    PM::Target: APeerManager<Descriptor = DynamicSocketDescriptor>,
+{
+    let ws_next = if *emit_read_events {
+        Either::Left(ws.next())
+    } else {
+        Either::Right(future::pending())
+    };
+
+    tokio::select! {
+        task_msg = task_rx.recv() => match task_msg.context("rust-lighting SocketDescriptor dropped")? {
+            BgTaskMessage::SendData { data, resume_read } => {
+                if resume_read {
+                    *emit_read_events = true;
+                }
+
+                ws.send(Message::Binary(data)).await?;
+            },
+            BgTaskMessage::Close => {
+                let _ = ws.close(None).await;
+                return Ok(ControlFlow::Break(()))
+            },
+        },
+        ws_msg = ws_next => {
+            let data = ws_msg.context("WS returned no data")??.into_data();
+            if let Ok(true) = peer_manager.as_ref().read_event(descriptor, &data) {
+                *emit_read_events = false; // Pause readign
+            }
+
+            peer_manager.as_ref().process_events();
+        }
+    }
+
+    Ok(ControlFlow::Continue(()))
+}
+
+enum BgTaskMessage {
+    SendData { data: Vec<u8>, resume_read: bool },
+    Close,
+}
+
+#[derive(Clone)]
+pub struct SocketDescriptor {
+    tx: mpsc::UnboundedSender<BgTaskMessage>,
+    id: u64,
+}
+
+impl Eq for SocketDescriptor {}
+impl PartialEq for SocketDescriptor {
+    fn eq(&self, o: &Self) -> bool {
+        self.id == o.id
+    }
+}
+
+impl Hash for SocketDescriptor {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.id.hash(state);
+    }
+}
+
+impl peer_handler::SocketDescriptor for SocketDescriptor {
+    fn send_data(&mut self, data: &[u8], resume_read: bool) -> usize {
+        // This isn't so great as we should be waiting for this to be sent before returning the
+        // amount of data sent (which implies that the send operation is done). This is so that the
+        // backpressure stuff works properly
```

Comment:
Just calling this out as I did in the description so it's easier to find

Reply:
Do you have any idea how we could tackle this limitation? Can we just add another channel so that the other task can let us know when the data has been sent?

Reply:
It's a little more complicated than that. If we don't send all the data when calling `send_data` then there is a function we need to call of the peer manager which results in `send_data` being called again. At first glance you'd think you can just make the 2nd a no-op, but there could be more data that's waiting to be sent by then. Therefore we need to keep track of how much we promised to send and how much extra must be sent.

Definitely possible to do this, just more work

Reply:
Thanks for explaining! As discussed, we don't have to change this until we run into problems.